sklearn/neighbors/_binary_tree.pxi

```diff
@@ -1055,10 +1055,11 @@ cdef class BinaryTree:
         if leaf_size < 1:
             raise ValueError("leaf_size must be greater than or equal to 1")
 
+        self.data_arr = check_array(data, dtype=DTYPE, order='C')
```

Comment:
We should move this validation on the top of `__init__` and change all `data.xxx` to `self.data_arr` to be sure that we use the validate array which will be a NumPy array.